server/shared/src/test/scala/org/http4s/server/middleware/CSRFSuite.scala

```diff
@@ -293,20 +290,38 @@ class CSRFSuite extends Http4sSuite {
   }
 
   test("not return the same token to mitigate BREACH") {
-    for {
+    val program = for {
       csrf <- csrfIO
       token <- csrf.generateToken[IO]
       raw1 <- csrf.extractRaw[IO](unlift(token)).flatMap(IO.fromEither)
-      res <- csrf.validate()(dummyRoutes)(
-        dummyRequest
-          .putHeaders(headerName.toString -> unlift(token))
-          .addCookie(cookieName, unlift(token))
-      )
-      rawContent = res.cookies.find(_.name == cookieName).map(_.content).getOrElse("")
-      raw2 <- csrf.extractRaw[IO](rawContent).flatMap(IO.fromEither)
-    } yield {
-      assertNotEquals(rawContent, token.toString)
-      assertEquals(raw1, raw2)
+    } yield (csrf, token, raw1)
+
+    TestControl.execute(program).flatMap { control =>
+      for {
+        _ <- control.advanceAndTick(1.millis)
+        res <-
+          control.results.flatMap(_.get match {
```

Comment:
It seems like `TestControl` behaves differently for [JS](https://github.com/http4s/http4s/actions/runs/5296953858/jobs/9588410555?pr=7166#step:10:3111) or maybe I'm doing something wrong 🤔  cc @armanbilge

Reply:
No, `TestControl` should behave the same. Does the code make some side-effecting call somewhere? e.g. something that relies on `delay(...)`, `blocking(...)` or `async(...)`?

Reply:
Aha, yes it does. It needs HMAC cryptography. This is async on JS (i.e. it takes real time to execute) so it cannot be used on the test runtime with mock time.

Reply:
Maybe we can just disable this test on JS, like this.
https://github.com/http4s/http4s/blob/88f6688aa63e26ac102773a754ea83b5c6bf06b5/tests/shared/src/test/scala/org/http4s/UriSuite.scala#L998-L999

Reply:
> This is async on JS (i.e. it takes real time to execute) so it cannot be used on the test runtime with mock time.

This is also worth noting somewhere on CE docs. It's a good-to-know thing!

Reply:
This is already mentioned in a couple places but should be noted more prominently.

This caveat si not specific to JS, it's specific to mixing `async` with `TestControl`. The reason it affects only JS here is because JS is using `async` for crypto but JVM and Native are using `delay`.

https://github.com/typelevel/cats-effect/blob/8ba8916407eb8f35fbffae76f72dac48d2dc7357/testkit/shared/src/main/scala/cats/effect/testkit/TestControl.scala#L209-L210

Reply:
I see, thanks. However, it's more about deadlock, rather than just getting different behaviour for JS. I mean, we didn't get any of the tests hanged forever. They were executed but without mocking the time.

Reply:
> They were executed but without mocking the time.

No, they did mock the time :) it just deadlock

> I mean, we didn't get any of the tests hanged forever.

It's because in `TestControl` deadlock does not mean the tests hang forever. Deadlock is detected as no result even though there are no fibers to run.

I realize this is all very confusing :) if you use `executeEmbed` as I recommended you would get a more informative error.

https://github.com/typelevel/cats-effect/blob/8ba8916407eb8f35fbffae76f72dac48d2dc7357/testkit/shared/src/main/scala/cats/effect/testkit/TestControl.scala#L413

Reply:
It's baffling why we don't get the same errors when ticking then (if an `IO` is deadlocked) 🤔

Reply:
Because when using the advanced API, it's not an error to get no result. The advance API simply lets check if a result is present or not, which is why it returns an `Option`. If you use `.get` on the `Option`, then you may get a surprising error.

https://github.com/http4s/http4s/blob/8bc0fee4887223fb33410cc6ec2ca1a527cd0b93/server/shared/src/test/scala/org/http4s/server/middleware/CSRFSuite.scala#L166

Reply:
Ah, gotcha